pkg/controlplane/control/manager.go

```diff
@@ -243,31 +336,96 @@ func (m *Manager) AddImport(ctx context.Context, imp *v1alpha1.Import) (err erro
 		return err
 	}
 
-	if imp.Namespace == m.namespace {
-		return nil
+	if imp.Namespace != m.namespace {
+		userService := &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      imp.Name,
+				Namespace: imp.Namespace,
+				Labels:    make(map[string]string),
+			},
+			Spec: v1.ServiceSpec{
+				ExternalName: fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, m.namespace),
+				Type:         v1.ServiceTypeExternalName,
+			},
+		}
+
+		if err := m.addImportService(ctx, imp, userService); err != nil {
+			return err
+		}
 	}
 
-	userService := &v1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      imp.Name,
-			Namespace: imp.Namespace,
-			Labels:    make(map[string]string),
-		},
-		Spec: v1.ServiceSpec{
-			ExternalName: fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, m.namespace),
-			Type:         v1.ServiceTypeExternalName,
-		},
+	if imp.Spec.DnsName != "" {
+		if err := addCoreDnsRewrite(ctx, m, &importName, imp.Spec.DnsName); err != nil {
+			m.logger.Errorf("Failed to configure CoreDns: %v.", err)
+			return err
+		}
+	}
+	return nil
+
+}
+
+// Remove coredns rewrite for a given external dns service
+func removeCoreDnsRewrite(ctx context.Context, m *Manager, name *types.NamespacedName) error {
+	corednsName := types.NamespacedName{
+		Name:      "coredns",
+		Namespace: "kube-system",
 	}
+	var cm v1.ConfigMap
 
-	return m.addImportService(ctx, imp, userService)
+	if err := m.client.Get(ctx, corednsName, &cm); err != nil {
+		if k8serrors.IsNotFound(err) {
+			m.logger.Warnf("coredns configmap not found.")
+			return nil
+		} else {
+			return err
+		}
+	}
+	if data, ok := cm.Data["Corefile"]; ok {
+		// remove trailing end-of-line
+		dataEol := strings.TrimSuffix(data, "\n")
+		// break into lines
+		lines := strings.Split(dataEol, "\n")
+		serviceFqdn := fmt.Sprintf("%s.%s.svc.cluster.local", name.Name, name.Namespace)
+
+		var coreFileUpdated = false
+		for i, line := range lines {
+			if strings.Contains(line, serviceFqdn) {
+				// remove matched line
+				lines = append(lines[:i], lines[i+1:]...)
+				coreFileUpdated = true
+				break
+			}
+		}
+
+		if coreFileUpdated {
+			// update configmap and restart the pods
+			var newLines string = ""
+			for _, line := range lines {
+				// return back EOL
+				newLines += (line + "\n")
+			}
+			cm.Data["Corefile"] = newLines
+			if err := m.client.Update(ctx, &cm); err != nil {
+				return err
+			}
+
+			if err := coreDnsRestart(ctx, m); err != nil {
+				return err
+			}
+		}
+	} else {
+		return errors.New("coredns configmap['Corefile'] not found")
+	}
+
+	return nil
 }
 
 // DeleteImport removes the listening socket of a previously imported service.
 func (m *Manager) DeleteImport(ctx context.Context, name types.NamespacedName) error {
 	m.logger.Infof("Deleting import '%s/%s'.", name.Namespace, name.Name)
 
 	// delete user service
-	errs := make([]error, 3)
+	errs := make([]error, 4)
```

Comment:
nit: I realize it is not part of the new code, but I wonder if using append won't be a cleaner solution instead of assigning specific array entries. Unless, of course, there is a reliance on the specific order even when there are nil errors in the array.
Alternately, this might be better served by `errors.Join()`.
@kfirtoledo @orozery - not familiar with this code well enough to say. Please weigh in.